Kernel/VM/Space.cpp

```diff
@@ -158,6 +158,63 @@ Region* Space::find_region_containing(const Range& range)
     return nullptr;
 }
 
+Vector<Region*> Space::find_regions_from_range(Range* range)
+{
+    Vector<Region*> regions = {};
+
+    ScopedSpinLock lock(m_lock);
+
+    // FIXME: Maybe take the cache from the single lookup?
+
+    size_t size = page_round_up(range->size());
+
+    bool found_one = true;
+    while (range && found_one) {
+        found_one = false;
+
+        for (auto& region : m_regions) {
+            if (region.vaddr() == range->base())
+                if (region.size() == size) {
+                    regions.append(&region);
+
+                    auto split_range = range->carve(region.range());
+                    range = split_range.is_empty() ? &(split_range.last()) : nullptr;
+
+                    // FIXME: Is it guaranteed that regions are in order?
+                    // If yes than we do not need to break here
+                    found_one = true;
+                    break;
+                } else
+                    return {};
+        }
+    }
+    return regions;
+}
+Vector<Region*> Space::find_regions_containing(Range* range)
```

Comment:
This duplicates a lot of code. Is this really necessary?

Reply:
maybe, maybe not it was the one thing that came to my mind looking at the old code, surely you could try to break it out into a lot of samller funtions, but that would be spaghtti code.

If you have another idea, how to solve this I'd like to here it, I'm not exactly a CS student..

Reply:
The implementation should now be shorter and as you have discibed in your comment

Reply:
Why not just a `for_each_intersected` template function that just iterates over them and only handles the intersecting regions?

Reply:
because with that we could not check if we bailed out with EPERM or if there actually were some regions to unmap

Reply:
Not following. Just capture a local variable by reference in the iteration lambda?

Reply:
Okay ive changed it to a lambda based aproach, though I don't think the code readability nor the performance of that code really changed thourgh that